File: erpnext/accounts/report/account_balance/account_balance.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals

import frappe
from frappe import _

from erpnext.accounts.utils import get_balance_on


def execute(filters=None):
	filters = frappe._dict(filters or {})
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data

def get_columns(filters):
	columns = [
		{
			"label": _("Account"),
			"fieldtype": "Link",
			"fieldname": "account",
			"options": "Account",
			"width": 100
		},
		{
			"label": _("Currency"),
			"fieldtype": "Link",
			"fieldname": "currency",
			"options": "Currency",
			"hidden": 1,
			"width": 50
		},
		{
			"label": _("Balance"),
			"fieldtype": "Currency",
			"fieldname": "balance",
			"options": "currency",
			"width": 100
		}
	]

	return columns

def get_conditions(filters):
	conditions = {}

	if filters.account_type:
		conditions["account_type"] = filters.account_type
		return conditions

	if filters.company:
		conditions["company"] = filters.company

	if filters.root_type:
		conditions["root_type"] = filters.root_type

	return conditions

def get_data(filters):

	data = []
	conditions = get_conditions(filters)
	accounts = frappe.db.get_all("Account", fields=["name", "account_currency"],
		filters=conditions, order_by='name')

	for d in accounts:
		balance = get_balance_on(d.name, date=filters.report_date)
		row = {"account": d.name, "balance": balance, "currency": d.account_currency}

		data.append(row)

	return data

File: erpnext/accounts/report/account_balance/account_balance.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Account Balance"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "report_date",
			label: __("Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "root_type",
			label: __("Root Type"),
			fieldtype: "Select",
			options: [
				{ value: "Asset", label: __("Asset") },
				{ value: "Liability", label: __("Liability") },
				{ value: "Income", label: __("Income") },
				{ value: "Expense", label: __("Expense") },
				{ value: "Equity", label: __("Equity") },
			],
		},
		{
			fieldname: "account_type",
			label: __("Account Type"),
			fieldtype: "Select",
			options: [
				{ value: "Accumulated Depreciation", label: __("Accumulated Depreciation") },
				{ value: "Asset Received But Not Billed", label: __("Asset Received But Not Billed") },
				{ value: "Bank", label: __("Bank") },
				{ value: "Cash", label: __("Cash") },
				{ value: "Chargeble", label: __("Chargeble") },
				{ value: "Capital Work in Progress", label: __("Capital Work in Progress") },
				{ value: "Cost of Goods Sold", label: __("Cost of Goods Sold") },
				{ value: "Depreciation", label: __("Depreciation") },
				{ value: "Equity", label: __("Equity") },
				{ value: "Expense Account", label: __("Expense Account") },
				{
					value: "Expenses Included In Asset Valuation",
					label: __("Expenses Included In Asset Valuation"),
				},
				{ value: "Expenses Included In Valuation", label: __("Expenses Included In Valuation") },
				{ value: "Fixed Asset", label: __("Fixed Asset") },
				{ value: "Income Account", label: __("Income Account") },
				{ value: "Payable", label: __("Payable") },
				{ value: "Receivable", label: __("Receivable") },
				{ value: "Round Off", label: __("Round Off") },
				{ value: "Stock", label: __("Stock") },
				{ value: "Stock Adjustment", label: __("Stock Adjustment") },
				{ value: "Stock Received But Not Billed", label: __("Stock Received But Not Billed") },
				{ value: "Tax", label: __("Tax") },
				{ value: "Temporary", label: __("Temporary") },
			],
		},
	],
};
